io/plugins/eu.esdihumboldt.hale.io.csv/src/eu/esdihumboldt/hale/io/csv/writer/internal/CSVInstanceWriter.java

```diff
@@ -76,27 +79,71 @@ public boolean isCancelable() {
 	protected IOReport execute(ProgressIndicator progress, IOReporter reporter)
 			throws IOProviderConfigurationException, IOException {
 
-		boolean solveNestedProperties = getParameter(
-				InstanceTableIOConstants.SOLVE_NESTED_PROPERTIES).as(Boolean.class, false);
-		// XXX what does "solve nested properties" mean?
-
 		// get separation, quote and escape sign
 		sep = CSVUtil.getSep(this);
 		quote = CSVUtil.getQuote(this);
 		esc = CSVUtil.getEscape(this);
 
-		List<String> headerRow = new ArrayList<String>(); // empty list
-
-		// get the parameter to get the type definition
 		String exportType = getParameter(InstanceTableIOConstants.EXPORT_TYPE).as(String.class);
-		QName selectedTypeName = null;
+		ArrayList<QName> selectedFeatureTypes = new ArrayList<>();
+
+		if (exportType != null) {
+			String[] splitExportType = exportType.split(",");
+			for (String featureType : splitExportType) {
+				selectedFeatureTypes.add(QName.valueOf(featureType));
+			}
 
-		if (exportType != null && !exportType.equals("") && !exportType.equals(" ")) {
-			selectedTypeName = QName.valueOf(exportType);
+			for (QName selectedTypeName : selectedFeatureTypes) {
+				// get all instances of the selected Type
+				InstanceCollection instances = getInstanceCollection(selectedTypeName);
+
+				if (selectedFeatureTypes.size() != 1 && getTarget().getLocation() != null) {
+					URI location = getTarget().getLocation();
+
+					String filePath = Paths.get(location).getParent().toString();
+					String baseFilename = Paths.get(location).getFileName().toString();
+					baseFilename = baseFilename.substring(0, baseFilename.lastIndexOf("."));
+					String filenameWithType = filePath + FileSystems.getDefault().getSeparator()
+							+ baseFilename + InstanceTableIOConstants.UNDERSCORE
+							+ selectedTypeName.getLocalPart()
+							+ InstanceTableIOConstants.CSV_EXTENSION;
+
+					File outputFile = new File(filenameWithType);
+					try {
+						FileOutputStream fos = new FileOutputStream(outputFile);
+						addCSVFileByQName(reporter, fos, instances);
+						fos.close();
```

Comment:
The stream should also be closed when an error occurs. Usually this is done in a `finally` block, or even better in a try-with-resources as mentioned earlier:

> For resources like streams you should always make sure to properly close them. If this is not done this can for instance lead to open file handles that may prevent deleting such a file.
> 
> You could for instance use the [try-with-resources](https://www.baeldung.com/java-try-with-resources) that was introduced in Java 7.

Reply:
You can omit the explicit call to `close()` since that is done by the try-with-resources block automatically.

Reply:
I  know that is optional to have it or not. I can remove it.